scipy/optimize/_optimize.py

```diff
@@ -2025,7 +2025,7 @@ def terminate(warnflag, msg):
     cg_maxiter = 20*len(x0)
 
     xtol = len(x0) * avextol
-    update_l1norm = 2 * xtol
+    update_l1norm = np.finfo(float).max
```

Comment:
I cannot imagine this being the right solution to be honest. This means the `update_l1norm > xtol` clause of the while-loop will ~always be true, and completely ignores that this should have something to do with the initial norm.

The _very_ first thing here should be to fix the `UnboundLocalError`, not hide it from happening.

Reply:
I agree I think the correct fix is updating `update` -> `update_l1norm` on line 2119

Reply:
There are two related issues that I was trying to address with this:

1. UnboundLocalError if xtol=0
2. The function will not succeed if xtol=0, even if an exact solution can be found (as is the case with the example given in the bug (https://github.com/scipy/scipy/issues/20214)

I completely agree that the most important issue to fix is the Unbound Error, but if we only make the change `update` -> `update_l1norm` we will still fail to find the valid solution for the presented test case.

I could well be missing something, but I don't think the initial value _does_ need to be related to the initial norm. The variable is only used to enter the loop (which is guaranteed for any value that isn't zero). It is then updated at the end of the loop, it only appears to be used to track the exit condition of the while.

My idea with using the max float as the initial value is that it retains the guarantee of running at least once, and also stops any issues with people passing in extremely large values for `xtol` that could overflow.

I'd certainly welcome your thoughts @h-vetinari  and @j-bowhay - I'm very new to SciPy so could well be misunderstanding how this works.


Whatever the right solution, I should have added a test! I'm happy to add one in as suggested.

Reply:
Ok yes I see your point the initial assignment is effectively only to enter the loop (basically a do while). However I am not sure the change is correct because previously `2 * xtol` is always greater than `xtol` whereas now this not the case. Prehaps it should initially just be set to `np.inf`.

Regardless, the mistake on 2119 should be fixed and it would be good add the test.

Reply:
Ha sorry a very silly mistake from me, I miss read this as min not max, I think this should be fine

Reply:
Thanks. The latest revision also includes a test and corrects the mistake on line 2119, so I think that covers everything.

Reply:
There's one thing that is being changed here, which is that previously, `update_l1norm` maintained the shape of `x0`, and in particular, `xtol` will _also_ have the shape of `x0`.

I don't think this ever worked for `len(x0)>1` though, as `while update_l1norm > xtol:` would then run into:
```
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

I don't know the code here well enough, but this stands out to me...